server/etcdserver/server.go

```diff
@@ -316,6 +320,9 @@ func (bh *backendHooks) OnPreCommitUnsafe(tx backend.BatchTx) {
 		// save bh.confState
 		bh.confStateDirty = false
 	}
+	bh.versionStored.Do(func() {
+		sv.UnsafeUpdateStorageVersion(tx, version.Version)
```

Comment:
I still think that we should create a helper function (possibly in `etcdserver/version` package): 

version.UpdateVersion(tx) error {
  if (tx.Get(Meta, ConfState)) != nil {
      return err.Printf("ConfState is required for etcd-3.5");
  }
  // The same for Term. 
  
  sv.UnsafeUpdateStorageVersion(tx, version.Version)
}

I'm aware we will not backfill the whole history (yet)  but I think we need to start maintaining style, that makes us conscious about storage changes going forward.

Reply:
Problem is that those fields are not populated during bootstrap not before first commit. As Etcd doesn't distinguish if loaded db was just created or not, we won't be able to set a version on fresh instances this way. I tried rewriting code to start populating them, but it broke some e2e tests and code got pretty complicated.

Versioning backed schema is definitely on my mind as it would be important for upgrades and downgrades, but it requires much more work than just solving this problem. Our current goal is to be able to answer simple question, "what etcd version should I use to read a particular snapshot". For that we don't need schema, we just need to store version of etcd that created it. It should be able to eventually evolve to schema management.

I would propose to instead of storing storageVersion that needs to be verified and managed, to store a version of etcd that created snapshot. It could be called "authorVersion" and instead of being set during bootstrap it would be set during snapshot creation. This way we solve the original problem and avoid creating half backed storage versioning.

WDYT?

Reply:
> Our current goal is to be able to answer simple question, "what etcd version should I use to read a particular snapshot". 

To me the question is exactly about the schema and not about 'about what version to wrote it'.

The problem is that developers of version N+2 of etcd does not remember that some fields might be unset in version N bbolt field.  If we have version N,  upgrade to N+1, run it for 10s but not fix the fields and then N+2, we are likely to either: 
  - experience outage
  - or keep all over the code safety checks what to do if the data does not exists (so not pay the technical depth). 

TBH the information about which version wrote the file is misleading, as ability to read the files is driven by version of bbolt (and is generally compatible within bbolt v1). What matters is bbolt (&wal) schema.  


What would happen if you moved the 'update' code post first snapshot ? It should be the point where all the properties are 'known' ?